# 🔍 Multimodal VQA System — Model Inference & Evaluation
**Member 02 | Phase 2: Model Loading, Inference Pipeline, and Evaluation**

## Overview
This notebook demonstrates the complete Member 02 pipeline:
- Loading BLIP-VQA model with caching
- Running inference on 48 curated VQA test

## Step 1 — Repository Setup

In [1]:
!git clone https://github.com/hasana157/image-insight-VQA.git
%cd image-insight-VQA


Cloning into 'image-insight-VQA'...
remote: Enumerating objects: 125, done.
remote: Counting objects: 100% (125/125), done.
remote: Compressing objects: 100% (103/103), done.
remote: Total 125 (delta 29), reused 110 (delta 17), pack-reused 0 (from 0)
Receiving objects: 100% (125/125), 5.28 MiB | 38.87 MiB/s, done.
Resolving deltas: 100% (29/29), done.
/content/image-insight-VQA


## Step 2 — Environment Setup

In [2]:
!pip install torch torchvision transformers Pillow pandas matplotlib -q

In [3]:
import torch
print(torch.cuda.is_available())  # Should print True
print(torch.cuda.get_device_name(0))  # Should print something like Tesla T4

True
Tesla T4


In [4]:
!pip install transformers accelerate -q

## Step 3 — Model Loading

In [5]:
from src.model_loader import load_model

processor, model = load_model()
print("Model loaded successfully!")

[model_loader] Loading model: Salesforce/blip-vqa-base on cuda ...


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


preprocessor_config.json:   0%|          | 0.00/445 [00:00<?, ?B/s]

The image processor of type `BlipImageProcessor` is now loaded as a fast processor by default, even if the model checkpoint was saved with a slow processor. This is a breaking change and may produce slightly different outputs. To continue using the slow processor, instantiate this class with `use_fast=False`. 


config.json: 0.00B [00:00, ?B/s]

tokenizer_config.json:   0%|          | 0.00/592 [00:00<?, ?B/s]

vocab.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

special_tokens_map.json:   0%|          | 0.00/125 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.54G [00:00<?, ?B/s]

Loading weights:   0%|          | 0/788 [00:00<?, ?it/s]

The tied weights mapping and config for this model specifies to tie text_decoder.bert.embeddings.word_embeddings.weight to text_decoder.cls.predictions.decoder.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning
BlipForQuestionAnswering LOAD REPORT from: Salesforce/blip-vqa-base
Key                                       | Status     |  | 
------------------------------------------+------------+--+-
text_encoder.embeddings.position_ids      | UNEXPECTED |  | 
text_decoder.bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


[model_loader] Model ready in 29.0s
Model loaded successfully!


## Step 4 — Single Inference Test

In [12]:
import importlib
import sys

# Remove cached modules
for mod in list(sys.modules.keys()):
    if "src" in mod:
        del sys.modules[mod]

# Now reimport
from src.inference import answer_question
print("Import successful!")

Import successful!


In [13]:
import os
images = os.listdir("data/sample_images")
print(images[:5])

['COCO_val2014_000000524436.jpg', 'COCO_val2014_000000393225.jpg', 'COCO_val2014_000000393282.jpg', 'COCO_val2014_000000393288.jpg', 'COCO_val2014_000000262200.jpg']


## Step 5 — Batch Inference (48 rows)

In [14]:
answer, inf_time = answer_question(
    "data/sample_images/COCO_val2014_000000524436.jpg",
    "What is in the image?"
)
print(f"Answer: {answer}")
print(f"Time:   {inf_time}s")

[model_loader] Loading model: Salesforce/blip-vqa-base on cuda ...


Loading weights:   0%|          | 0/788 [00:00<?, ?it/s]

The tied weights mapping and config for this model specifies to tie text_decoder.bert.embeddings.word_embeddings.weight to text_decoder.cls.predictions.decoder.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning
BlipForQuestionAnswering LOAD REPORT from: Salesforce/blip-vqa-base
Key                                       | Status     |  | 
------------------------------------------+------------+--+-
text_encoder.embeddings.position_ids      | UNEXPECTED |  | 
text_decoder.bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


[model_loader] Model ready in 3.6s
Answer: people
Time:   1.9921s


In [15]:
from src.inference import run_batch_inference

results = run_batch_inference("data/vqa_test_set.csv")

[inference] Running batch inference on 48 rows...
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [01] ✓ Q: Where is he looking? | Pred: down | GT: down | 0.2057s
[model_loader] Using cached model: Salesforce/blip-vqa-base


/content/image-insight-VQA/src/inference.py:171: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'down' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.at[i, "predicted_answer"]  = predicted


  [02] ✓ Q: Is this a creamy soup? | Pred: no | GT: no | 0.1821s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [03] ✓ Q: Is this rice noodle soup? | Pred: yes | GT: yes | 0.1643s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [04] ✓ Q: What is to the right of the soup? | Pred: chopsticks | GT: chopsticks | 0.1538s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [05] ✓ Q: What does the truck on the left sell? | Pred: ice cream | GT: ice cream | 0.1229s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [06] ✓ Q: Is it daylight in this picture? | Pred: yes | GT: yes | 0.1083s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [07] ✓ Q: How many are playing ball? | Pred: 1 | GT: 1 | 0.1056s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [08] ✓ Q: Is there a chain link fence in the image? | Pred: no | GT: no | 0.0921s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [09] ✓ Q: Is the boy playing b

## Step 6 — Evaluation before Normalization

In [16]:
from src.evaluation import evaluate_predictions

summary = evaluate_predictions("data/predictions.csv")

[evaluation] Metrics saved to results/metrics_summary.csv
[evaluation] 5 failure cases saved to results/failure_cases.csv
[evaluation] Chart saved to results/question_type_accuracy.png
[evaluation] Chart saved to results/inference_time_chart.png

  EVALUATION SUMMARY
  Total rows:          48
  Overall accuracy:    89.6%
  Failure rate:        10.4%
  Avg inference time:  0.1002s

  Per-type accuracy:
    action           7/8  (87.5%)
    color            7/8  (87.5%)
    counting         6/8  (75.0%)
    object           7/8  (87.5%)
    spatial_scene    8/8  (100.0%)
    yes_no           8/8  (100.0%)



## Step 7 — Normalization

In [18]:
synonym_patch = '''

# Synonym map for answer normalization
ANSWER_SYNONYMS = {
    "bikes": "bicycles",
    "bicycle": "bicycles",
    "motorbike": "motorcycle",
    "motorbikes": "motorcycles",
    "sofa": "couch",
    "settee": "couch",
    "automobile": "car",
    "auto": "car",
    "cab": "taxi",
    "plane": "airplane",
    "aeroplane": "airplane",
    "tv": "television",
    "telly": "television",
    "fridge": "refrigerator",
    "specs": "glasses",
    "spectacles": "glasses",
    "jumping": "riding",  # horse context
    "leaping": "jumping",
}
'''

with open("src/config.py", "a") as f:
    f.write(synonym_patch)

print("Synonym map added to config.py")

Synonym map added to config.py


In [19]:
new_normalize = '''
def normalize_answer(answer: str, question: str = "") -> str:
    from src.config import ANSWER_SYNONYMS
    if not answer:
        return "unknown"
    answer = answer.strip().lower().rstrip(".,!?;:")
    yes_variants = {"yes", "yeah", "yep", "yup", "correct", "true", "right"}
    no_variants  = {"no", "nope", "nah", "false", "wrong", "not"}
    if answer in yes_variants:
        return "yes"
    if answer in no_variants:
        return "no"
    for article in ("a ", "an ", "the "):
        if answer.startswith(article):
            answer = answer[len(article):]
            break
    # Apply synonym normalization
    answer = ANSWER_SYNONYMS.get(answer, answer)
    return answer
'''

# Read inference.py
with open("src/inference.py", "r") as f:
    content = f.read()

# Find and replace old normalize_answer function
import re
content = re.sub(
    r'def normalize_answer\(.*?(?=\ndef |\Z)',
    new_normalize.strip() + "\n\n",
    content,
    flags=re.DOTALL
)

with open("src/inference.py", "w") as f:
    f.write(content)

print("normalize_answer() patched in inference.py")

normalize_answer() patched in inference.py


In [21]:
import sys
for mod in list(sys.modules.keys()):
    if "src" in mod:
        del sys.modules[mod]

# Read config
with open("src/config.py", "r") as f:
    content = f.read()

# Fix the direction
content = content.replace('"bikes": "bicycles"', '"bicycles": "bikes"')
content = content.replace('"bicycle": "bicycles"', '"bicycle": "bikes"')

with open("src/config.py", "w") as f:
    f.write(content)

print("Synonym direction fixed")

Synonym direction fixed


## Step 8 — Batch Inference (48 rows)

In [22]:
for mod in list(sys.modules.keys()):
    if "src" in mod:
        del sys.modules[mod]

from src.inference import run_batch_inference
results = run_batch_inference("data/vqa_test_set.csv")

[inference] Running batch inference on 48 rows...
[model_loader] Loading model: Salesforce/blip-vqa-base on cuda ...


Loading weights:   0%|          | 0/788 [00:00<?, ?it/s]

The tied weights mapping and config for this model specifies to tie text_decoder.bert.embeddings.word_embeddings.weight to text_decoder.cls.predictions.decoder.weight, but both are present in the checkpoints, so we will NOT tie them. You should update the config with `tie_word_embeddings=False` to silence this warning
BlipForQuestionAnswering LOAD REPORT from: Salesforce/blip-vqa-base
Key                                       | Status     |  | 
------------------------------------------+------------+--+-
text_encoder.embeddings.position_ids      | UNEXPECTED |  | 
text_decoder.bert.embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


[model_loader] Model ready in 3.3s
  [01] ✓ Q: Where is he looking? | Pred: down | GT: down | 0.1914s
[model_loader] Using cached model: Salesforce/blip-vqa-base


/content/image-insight-VQA/src/inference.py:149: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'down' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  


  [02] ✓ Q: Is this a creamy soup? | Pred: no | GT: no | 0.1752s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [03] ✓ Q: Is this rice noodle soup? | Pred: yes | GT: yes | 0.1775s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [04] ✓ Q: What is to the right of the soup? | Pred: chopsticks | GT: chopsticks | 0.2639s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [05] ✓ Q: What does the truck on the left sell? | Pred: ice cream | GT: ice cream | 0.2063s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [06] ✓ Q: Is it daylight in this picture? | Pred: yes | GT: yes | 0.1518s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [07] ✓ Q: How many are playing ball? | Pred: 1 | GT: 1 | 0.1575s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [08] ✓ Q: Is there a chain link fence in the image? | Pred: no | GT: no | 0.1496s
[model_loader] Using cached model: Salesforce/blip-vqa-base
  [09] ✓ Q: Is the boy playing b

## Step 9 — Evaluation Results

In [23]:
from src.evaluation import evaluate_predictions
summary = evaluate_predictions("data/predictions.csv")

[evaluation] Metrics saved to results/metrics_summary.csv
[evaluation] 3 failure cases saved to results/failure_cases.csv
[evaluation] Chart saved to results/question_type_accuracy.png
[evaluation] Chart saved to results/inference_time_chart.png

  EVALUATION SUMMARY
  Total rows:          48
  Overall accuracy:    93.8%
  Failure rate:        6.2%
  Avg inference time:  0.1225s

  Per-type accuracy:
    action           8/8  (100.0%)
    color            7/8  (87.5%)
    counting         6/8  (75.0%)
    object           8/8  (100.0%)
    spatial_scene    8/8  (100.0%)
    yes_no           8/8  (100.0%)



## Step 10 — Push to GitHub

In [29]:
!git pull origin main

remote: Enumerating objects: 7, done.
remote: Counting objects: 100% (7/7), done.
remote: Compressing objects: 100% (1/1), done.
remote: Total 4 (delta 3), reused 4 (delta 3), pack-reused 0 (from 0)
Unpacking objects: 100% (4/4), 567 bytes | 94.00 KiB/s, done.
From https://github.com/hasana157/image-insight-VQA
 * branch            main       -> FETCH_HEAD
   3bb9a15..69066a2  main       -> origin/main
Updating 3bb9a15..69066a2
Fast-forward
 src/config.py | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)


In [32]:
!git add data/predictions.csv results/ src/config.py src/inference.py
!git commit -m "Phase 2: final results - 93.8% accuracy, all evaluation outputs"
!git push origin main

[main d4a66a3] Phase 2: final results - 93.8% accuracy, all evaluation outputs
 7 files changed, 94 insertions(+), 28 deletions(-)
 create mode 100644 results/failure_cases.csv
 create mode 100644 results/inference_time_chart.png
 create mode 100644 results/question_type_accuracy.png
Enumerating objects: 20, done.
Counting objects: 100% (20/20), done.
Delta compression using up to 2 threads
Compressing objects: 100% (12/12), done.
Writing objects: 100% (12/12), 102.27 KiB | 11.36 MiB/s, done.
Total 12 (delta 6), reused 0 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (6/6), completed with 5 local objects.
To https://github.com/hasana157/image-insight-VQA.git
   69066a2..d4a66a3  main -> main


In [35]:
from google.colab import files
files.download("data/predictions.csv")
files.download("results/metrics_summary.csv")
files.download("results/failure_cases.csv")
files.download("results/question_type_accuracy.png")
files.download("results/inference_time_chart.png")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Summary
| Metric | Result |
|---|---|
| Total test cases | 48 |
| Overall accuracy | 93.8% |
| Average inference time | 0.1225s |
| action | 100% |
| yes_no | 100% |
| object | 100% |
| spatial_scene | 100% |
| color | 87.5% |
| counting | 75.0% |

### Failure Analysis
- **Counting error:** Predicted 3 giraffes, GT was 2 (partial occlusion)
- **Counting error:** Predicted 8 pictures, GT was 7 (cluttered scene)
- **Color confusion:** Predicted red church, GT was brown (lighting ambiguity)